File: apps/hwp-lab/src/lib/xlsxRoster.ts
// First-sheet .xlsx roster reader (issue #40).
// Contract shared with crates/auto-hwp-cli/src/xlsx_roster.rs — same honesty as CSV:
// extra sheets, merged cells, and values past the header are errors, not silent shifts.

export async function parseXlsxRoster(input: ArrayBuffer | Uint8Array): Promise<Record<string, string>[]> {
  const bytes = input instanceof Uint8Array ? input : new Uint8Array(input);
  const files = await unzip(bytes);
  const workbook = textPart(files, "xl/workbook.xml");
  if (!workbook) throw new Error("xlsx를 읽지 못했습니다: xl/workbook.xml 없음");
  const sheetIds = workbookSheetRids(workbook);
  if (!sheetIds.length) throw new Error("xlsx 시트가 없습니다");
  if (sheetIds.length !== 1) {
    throw new Error(
      `xlsx 시트가 ${sheetIds.length}개입니다 — 첫 시트만 지원합니다. 나머지 시트를 지우거나 JSON/CSV로 보내주세요`,
    );
  }
  const rels = textPart(files, "xl/_rels/workbook.xml.rels") ?? "";
  const relMap = parseRels(rels);
  const target = relMap.get(sheetIds[0]);
  if (!target) throw new Error(`xlsx를 읽지 못했습니다: 시트 관계 ${sheetIds[0]} 없음`);
  const sheetPath = resolveXlTarget(target);
  const sheet = textPart(files, sheetPath);
  if (!sheet) throw new Error(`xlsx를 읽지 못했습니다: ${sheetPath} 없음`);
  const sstXml = textPart(files, "xl/sharedStrings.xml");
  const sst = sstXml ? parseSharedStrings(sstXml) : [];
  return rowsFromSheet(sheet, sst);
}

function textPart(files: Map<string, Uint8Array>, name: string): string | null {
  const key = name.replace(/^\/+/, "");
  const data = files.get(key) ?? files.get(`/${key}`);
  if (!data) return null;
  let s = new TextDecoder("utf-8").decode(data);
  if (s.charCodeAt(0) === 0xfeff) s = s.slice(1);
  return s;
}

function localName(tag: string): string {
  const i = tag.indexOf(":");
  return i >= 0 ? tag.slice(i + 1) : tag;
}

function workbookSheetRids(xml: string): string[] {
  const ids: string[] = [];
  const re = /<([A-Za-z0-9:]*sheet)\b([^>]*)\/?>/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(xml))) {
    if (localName(m[1]) !== "sheet") continue;
    const rid = attr(m[2], "id");
    if (!rid) throw new Error("xlsx를 읽지 못했습니다: 시트 r:id 없음");
    ids.push(rid);
  }
  return ids;
}

function parseRels(xml: string): Map<string, string> {
  const map = new Map<string, string>();
  const re = /<([A-Za-z0-9:]*Relationship)\b([^>]*)\/?>/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(xml))) {
    if (localName(m[1]) !== "Relationship") continue;
    const id = attr(m[2], "Id");
    const target = attr(m[2], "Target");
    if (id && target) map.set(id, target);
  }
  return map;
}

function resolveXlTarget(target: string): string {
  const t = target.replace(/\\/g, "/");
  if (t.startsWith("/xl/")) return t.slice(1);
  if (t.startsWith("xl/")) return t;
  if (t.startsWith("/")) return `xl${t}`;
  return `xl/${t}`;
}

function parseSharedStrings(xml: string): string[] {
  const out: string[] = [];
  const siRe = /<([A-Za-z0-9:]*si)\b[^>]*>([\s\S]*?)<\/\1>/g;
  let m: RegExpExecArray | null;
  while ((m = siRe.exec(xml))) {
    const inner = m[2].replace(/<([A-Za-z0-9:]*rPh)\b[^>]*>[\s\S]*?<\/\1>/g, "");
    const texts = [...inner.matchAll(/<([A-Za-z0-9:]*t)\b[^>]*>([\s\S]*?)<\/\1>/g)].map((t) => decodeXml(t[2]));
    out.push(texts.join(""));
  }
  return out;
}

function rowsFromSheet(xml: string, sst: string[]): Record<string, string>[] {
  if (/<([A-Za-z0-9:]*mergeCell)\b/i.test(xml)) {
    throw new Error("xlsx에 병합 셀이 있습니다 — 병합을 풀거나 JSON으로 보내주세요");
  }
  const cells = new Map<string, string>();
  const cellRe = /<([A-Za-z0-9:]*c)\b([^>]*)>([\s\S]*?)<\/\1>|<([A-Za-z0-9:]*c)\b([^>]*)\/>/g;
  let m: RegExpExecArray | null;
  while ((m = cellRe.exec(xml))) {
    const attrs = m[2] ?? m[5] ?? "";
    const inner = m[3] ?? "";
    if (localName(m[1] ?? m[4] ?? "") !== "c") continue;
    const ref = attr(attrs, "r");
    if (!ref) continue;
    const kind = attr(attrs, "t") ?? "";
    const parsed = parseCellRef(ref);
    if (!parsed) continue;
    const vMatch = inner.match(/<([A-Za-z0-9:]*v)\b[^>]*>([\s\S]*?)<\/\1>/);
    const tMatch = inner.match(/<([A-Za-z0-9:]*is)\b[^>]*>[\s\S]*?<([A-Za-z0-9:]*t)\b[^>]*>([\s\S]*?)<\/\2>/);
    const raw = tMatch ? decodeXml(tMatch[3]) : vMatch ? decodeXml(vMatch[2]) : "";
    const value = decodeCell(kind, raw, sst);
    if (value) cells.set(`${parsed.row},${parsed.col}`, value);
  }
  return tableFromCells(cells);
}

function decodeCell(kind: string, raw: string, sst: string[]): string {
  if (kind === "s") {
    const i = Number(raw.trim());
    if (!Number.isInteger(i) || i < 0 || i >= sst.length) {
      throw new Error(`xlsx를 읽지 못했습니다: 공유 문자열 인덱스 ${raw}`);
    }
    return sst[i];
  }
  if (kind === "b") return raw.trim() === "1" ? "TRUE" : "FALSE";
  return raw.replace(/\s+$/g, "");
}

function tableFromCells(cells: Map<string, string>): Record<string, string>[] {
  if (!cells.size) throw new Error("xlsx 헤더가 비어 있습니다");
  let minRow = Infinity;
  let maxRow = 0;
  let maxHeader = 0;
  for (const key of cells.keys()) {
    const [r, c] = key.split(",").map(Number);
    if (r < minRow) minRow = r;
    if (r > maxRow) maxRow = r;
  }
  for (const key of cells.keys()) {
    const [r, c] = key.split(",").map(Number);
    if (r === minRow && c > maxHeader) maxHeader = c;
  }
  const headers: string[] = [];
  for (let col = 0; col <= maxHeader; col++) {
    const name = (cells.get(`${minRow},${col}`) ?? "").trim();
    if (!name) {
      throw new Error(`xlsx 헤더 ${col + 1}열이 비어 있습니다 — 빈 헤더는 JSON으로 보내주세요`);
    }
    headers.push(name);
  }
  const rows: Record<string, string>[] = [];
  for (let row = minRow + 1; row <= maxRow; row++) {
    for (const key of cells.keys()) {
      const [r, c] = key.split(",").map(Number);
      if (r === row && c > maxHeader) {
        throw new Error(`xlsx ${row}행: 헤더 밖의 값이 있습니다 (${c + 1}열)`);
      }
    }
    const rec: Record<string, string> = {};
    let any = false;
    headers.forEach((h, i) => {
      const val = cells.get(`${row},${i}`) ?? "";
      if (val) any = true;
      rec[h] = val;
    });
    if (any) rows.push(rec);
  }
  if (!rows.length) throw new Error("xlsx 데이터 행이 없습니다");
  return rows;
}

function parseCellRef(ref: string): { row: number; col: number } | null {
  const m = /^\$?([A-Za-z]+)\$?(\d+)$/.exec(ref.trim());
  if (!m) return null;
  const row = Number(m[2]);
  if (!row) return null;
  let col = 0;
  for (const ch of m[1].toUpperCase()) {
    col = col * 26 + (ch.charCodeAt(0) - 64);
  }
  return { row, col: col - 1 };
}

function attr(attrs: string, name: string): string | null {
  const re = new RegExp(`(?:^|\\s)(?:[A-Za-z0-9]+:)?${name}="([^"]*)"`);
  const m = re.exec(attrs);
  return m ? decodeXml(m[1]) : null;
}

function decodeXml(s: string): string {
  return s
    .replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, "$1")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&amp;/g, "&");
}

async function unzip(buf: Uint8Array): Promise<Map<string, Uint8Array>> {
  if (buf.length < 22 || buf[0] !== 0x50 || buf[1] !== 0x4b) {
    throw new Error("xlsx를 읽지 못했습니다: zip이 아닙니다");
  }
  const eocd = findEocd(buf);
  const entries = u16(buf, eocd + 10);
  let cd = u32(buf, eocd + 16);
  const files = new Map<string, Uint8Array>();
  for (let i = 0; i < entries; i++) {
    if (u32(buf, cd) !== 0x02014b50) throw new Error("xlsx를 읽지 못했습니다: zip 중앙 디렉터리");
    const method = u16(buf, cd + 10);
    const compSize = u32(buf, cd + 20);
    const nameLen = u16(buf, cd + 28);
    const extraLen = u16(buf, cd + 30);
    const commentLen = u16(buf, cd + 32);
    const localOff = u32(buf, cd + 42);
    const name = utf8(buf.subarray(cd + 46, cd + 46 + nameLen));
    const dataStart = localOff + 30 + u16(buf, localOff + 26) + u16(buf, localOff + 28);
    const comp = buf.subarray(dataStart, dataStart + compSize);
    files.set(name.replace(/^\/+/, ""), await inflate(method, comp));
    cd += 46 + nameLen + extraLen + commentLen;
  }
  return files;
}

function findEocd(buf: Uint8Array): number {
  for (let i = buf.length - 22; i >= 0; i--) {
    if (u32(buf, i) === 0x06054b50) return i;
  }
  throw new Error("xlsx를 읽지 못했습니다: zip EOCD 없음");
}

async function inflate(method: number, data: Uint8Array): Promise<Uint8Array> {
  if (method === 0) return data;
  if (method !== 8) throw new Error(`xlsx를 읽지 못했습니다: zip method ${method}`);
  const ds = new DecompressionStream("deflate-raw");
  const ab = await new Response(new Blob([data as BlobPart]).stream().pipeThrough(ds)).arrayBuffer();
  return new Uint8Array(ab);
}

function u16(b: Uint8Array, i: number): number {
  return b[i] | (b[i + 1] << 8);
}
function u32(b: Uint8Array, i: number): number {
  return (b[i] | (b[i + 1] << 8) | (b[i + 2] << 16) | (b[i + 3] << 24)) >>> 0;
}
function utf8(b: Uint8Array): string {
  return new TextDecoder("utf-8").decode(b);
}

File: crates/auto-hwp-cli/src/xlsx_roster.rs
//! First-sheet `.xlsx` roster reader (issue #40).
//!
//! Contract shared with `apps/hwp-lab/src/lib/xlsxRoster.ts`:
//! - exactly one worksheet; extra sheets are an error, not a silent sheet-0 pick
//! - row 1 is keys (한글 헤더 허용); later rows are documents
//! - trailing all-empty rows are ignored
//! - any merged cell is an error (no silent shift)
//! - a value past the last header column is an error
//!
//! Uses workspace `zip` + `quick-xml` only — no extra native runtime, no calamine.

use std::collections::BTreeMap;
use std::io::{Cursor, Read};
use std::path::Path;

use quick_xml::events::Event;
use quick_xml::Reader;
use serde_json::{Map, Value};

pub fn load_path(path: &Path) -> Result<Vec<Map<String, Value>>, String> {
    let bytes = std::fs::read(path).map_err(|e| format!("read {}: {e}", path.display()))?;
    parse_xlsx_roster(&bytes)
}

pub fn parse_xlsx_roster(bytes: &[u8]) -> Result<Vec<Map<String, Value>>, String> {
    let mut zip = zip::ZipArchive::new(Cursor::new(bytes.to_vec()))
        .map_err(|e| format!("xlsx를 읽지 못했습니다: {e}"))?;

    let workbook = read_zip(&mut zip, "xl/workbook.xml")?;
    let sheet_ids = workbook_sheet_rids(&workbook)?;
    if sheet_ids.is_empty() {
        return Err("xlsx 시트가 없습니다".into());
    }
    if sheet_ids.len() != 1 {
        return Err(format!(
            "xlsx 시트가 {}개입니다 — 첫 시트만 지원합니다. 나머지 시트를 지우거나 JSON/CSV로 보내주세요",
            sheet_ids.len()
        ));
    }

    let rels = read_zip(&mut zip, "xl/_rels/workbook.xml.rels").unwrap_or_default();
    let rel_map = parse_rels(&rels);
    let target = rel_map
        .get(&sheet_ids[0])
        .ok_or_else(|| format!("xlsx를 읽지 못했습니다: 시트 관계 {} 없음", sheet_ids[0]))?;
    let sheet_path = resolve_xl_target(target);

    let sst = read_zip(&mut zip, "xl/sharedStrings.xml")
        .ok()
        .map(|xml| parse_shared_strings(&xml))
        .transpose()?
        .unwrap_or_default();

    let sheet = read_zip(&mut zip, &sheet_path)?;
    rows_from_sheet(&sheet, &sst)
}

fn read_zip(zip: &mut zip::ZipArchive<Cursor<Vec<u8>>>, name: &str) -> Result<String, String> {
    let mut file = zip
        .by_name(name)
        .map_err(|_| format!("xlsx를 읽지 못했습니다: {name} 없음"))?;
    let mut buf = String::new();
    file.read_to_string(&mut buf)
        .map_err(|e| format!("xlsx를 읽지 못했습니다: {name}: {e}"))?;
    if buf.starts_with('\u{feff}') {
        buf = buf.trim_start_matches('\u{feff}').to_string();
    }
    Ok(buf)
}

fn local_name(q: &[u8]) -> &[u8] {
    match q.iter().rposition(|b| *b == b':') {
        Some(i) => &q[i + 1..],
        None => q,
    }
}

fn workbook_sheet_rids(xml: &str) -> Result<Vec<String>, String> {
    let mut reader = Reader::from_str(xml);
    reader.config_mut().trim_text(true);
    let mut buf = Vec::new();
    let mut ids = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Empty(e) | Event::Start(e)) => {
                let qn = e.name();
                if local_name(qn.as_ref()) == b"sheet" {
                    let mut rid = None;
                    for a in e.attributes().flatten() {
                        let key = local_name(a.key.as_ref());
                        if key == b"id" {
                            rid = Some(
                                a.decode_and_unescape_value(reader.decoder())
                                    .map_err(|err| err.to_string())?
                                    .into_owned(),
                            );
                        }
                    }
                    ids.push(rid.ok_or("xlsx를 읽지 못했습니다: 시트 r:id 없음")?);
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(format!("xlsx를 읽지 못했습니다: workbook.xml: {e}")),
            _ => {}
        }
        buf.clear();
    }
    Ok(ids)
}

fn parse_rels(xml: &str) -> BTreeMap<String, String> {
    let mut reader = Reader::from_str(xml);
    reader.config_mut().trim_text(true);
    let mut buf = Vec::new();
    let mut map = BTreeMap::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Empty(e) | Event::Start(e)) => {
                let qn = e.name();
                if local_name(qn.as_ref()) == b"Relationship" {
                    let mut id = None;
                    let mut target = None;
                    for a in e.attributes().flatten() {
                        let key = local_name(a.key.as_ref());
                        let val = a
                            .decode_and_unescape_value(reader.decoder())
                            .map(|v| v.into_owned())
                            .unwrap_or_default();
                        if key == b"Id" {
                            id = Some(val);
                        } else if key == b"Target" {
                            target = Some(val);
                        }
                    }
                    if let (Some(id), Some(target)) = (id, target) {
                        map.insert(id, target);
                    }
                }
            }
            Ok(Event::Eof) | Err(_) => break,
            _ => {}
        }
        buf.clear();
    }
    map
}

fn resolve_xl_target(target: &str) -> String {
    let t = target.replace('\\', "/");
    if t.starts_with("/xl/") {
        t.trim_start_matches('/').to_string()
    } else if t.starts_with("xl/") {
        t
    } else if t.starts_with('/') {
        format!("xl{}", t)
    } else {
        format!("xl/{t}")
    }
}

fn parse_shared_strings(xml: &str) -> Result<Vec<String>, String> {
    let mut reader = Reader::from_str(xml);
    reader.config_mut().trim_text(false);
    let mut buf = Vec::new();
    let mut out = Vec::new();
    let mut cur = String::new();
    let mut in_si = false;
    let mut in_t = false;
    let mut in_rph = false;
    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Start(e)) => {
                let qn = e.name();
                let name = local_name(qn.as_ref());
                if name == b"si" {
                    in_si = true;
                    cur.clear();
                } else if name == b"rPh" {
                    in_rph = true;
                } else if name == b"t" && in_si && !in_rph {
                    in_t = true;
                }
            }
            Ok(Event::Empty(e)) => {
                let qn = e.name();
                if local_name(qn.as_ref()) == b"si" {
                    out.push(String::new());
                }
            }
            Ok(Event::Text(t)) if in_t => {
                cur.push_str(&t.unescape().map_err(|e| e.to_string())?);
            }
            Ok(Event::CData(t)) if in_t => {
                cur.push_str(&String::from_utf8_lossy(&t.into_inner()));
            }
            Ok(Event::End(e)) => {
                let qn = e.name();
                let name = local_name(qn.as_ref());
                if name == b"t" {
                    in_t = false;
                } else if name == b"rPh" {
                    in_rph = false;
                } else if name == b"si" {
                    out.push(std::mem::take(&mut cur));
                    in_si = false;
                    in_t = false;
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(format!("xlsx를 읽지 못했습니다: sharedStrings.xml: {e}")),
            _ => {}
        }
        buf.clear();
    }
    Ok(out)
}

fn rows_from_sheet(xml: &str, sst: &[String]) -> Result<Vec<Map<String, Value>>, String> {
    let mut reader = Reader::from_str(xml);
    reader.config_mut().trim_text(false);
    let mut buf = Vec::new();

    let mut cells: BTreeMap<(u32, u32), String> = BTreeMap::new();
    let mut merges: Vec<(u32, u32, u32, u32)> = Vec::new();

    let mut cell_ref = String::new();
    let mut cell_type = String::new();
    let mut in_v = false;
    let mut in_t = false;
    let mut in_is = false;
    let mut text = String::new();

    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Start(e)) => {
                let qn = e.name();
                let name = local_name(qn.as_ref());
                if name == b"c" {
                    cell_ref.clear();
                    cell_type.clear();
                    text.clear();
                    in_is = false;
                    for a in e.attributes().flatten() {
                        let key = local_name(a.key.as_ref());
                        let val = a
                            .decode_and_unescape_value(reader.decoder())
                            .map(|v| v.into_owned())
                            .unwrap_or_default();
                        if key == b"r" {
                            cell_ref = val;
                        } else if key == b"t" {
                            cell_type = val;
                        }
                    }
                } else if name == b"v" {
                    in_v = true;
                    text.clear();
                } else if name == b"is" {
                    in_is = true;
                } else if name == b"t" && in_is {
                    in_t = true;
                } else if name == b"mergeCell" {
                    for a in e.attributes().flatten() {
                        if local_name(a.key.as_ref()) == b"ref" {
                            let val = a
                                .decode_and_unescape_value(reader.decoder())
                                .map_err(|err| err.to_string())?;
                            merges.push(parse_merge(&val)?);
                        }
                    }
                }
            }
            Ok(Event::Empty(e)) => {
                let qn = e.name();
                if local_name(qn.as_ref()) == b"mergeCell" {
                    for a in e.attributes().flatten() {
                        if local_name(a.key.as_ref()) == b"ref" {
                            let val = a
                                .decode_and_unescape_value(reader.decoder())
                                .map_err(|err| err.to_string())?;
                            merges.push(parse_merge(&val)?);
                        }
                    }
                }
            }
            Ok(Event::Text(t)) if in_v || in_t => {
                text.push_str(&t.unescape().map_err(|e| e.to_string())?);
            }
            Ok(Event::CData(t)) if in_v || in_t => {
                text.push_str(&String::from_utf8_lossy(&t.into_inner()));
            }
            Ok(Event::End(e)) => {
                let qn = e.name();
                let name = local_name(qn.as_ref());
                if name == b"v" {
                    in_v = false;
                } else if name == b"t" {
                    in_t = false;
                } else if name == b"is" {
                    in_is = false;
                } else if name == b"c" {
                    if !cell_ref.is_empty() {
                        if let Some((row, col)) = parse_cell_ref(&cell_ref) {
                            let value = decode_cell(&cell_type, &text, sst)?;
                            if !value.is_empty() {
                                cells.insert((row, col), value);
                            }
                        }
                    }
                    cell_ref.clear();
                    cell_type.clear();
                    text.clear();
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(format!("xlsx를 읽지 못했습니다: sheet: {e}")),
            _ => {}
        }
        buf.clear();
    }

    if !merges.is_empty() {
        return Err("xlsx에 병합 셀이 있습니다 — 병합을 풀거나 JSON으로 보내주세요".into());
    }

    table_from_cells(&cells)
}

fn decode_cell(kind: &str, raw: &str, sst: &[String]) -> Result<String, String> {
    match kind {
        "s" => {
            let i: usize = raw
                .trim()
                .parse()
                .map_err(|_| format!("xlsx를 읽지 못했습니다: 공유 문자열 인덱스 {raw}"))?;
            sst.get(i)
                .cloned()
                .ok_or_else(|| format!("xlsx를 읽지 못했습니다: 공유 문자열 {i} 없음"))
        }
        "b" => Ok(if raw.trim() == "1" {
            "TRUE".into()
        } else {
            "FALSE".into()
        }),
        "e" => Ok(raw.trim().to_string()),
        "inlineStr" | "str" | "" => Ok(raw.trim_end().to_string()),
        _ => Ok(raw.trim_end().to_string()),
    }
}

fn table_from_cells(
    cells: &BTreeMap<(u32, u32), String>,
) -> Result<Vec<Map<String, Value>>, String> {
    if cells.is_empty() {
        return Err("xlsx 헤더가 비어 있습니다".into());
    }
    let min_row = cells.keys().map(|(r, _)| *r).min().unwrap();
    let header_cols: Vec<u32> = cells
        .keys()
        .filter(|(r, _)| *r == min_row)
        .map(|(_, c)| *c)
        .collect();
    let max_header = *header_cols.iter().max().unwrap();
    let mut headers = Vec::new();
    for col in 0..=max_header {
        let Some(name) = cells.get(&(min_row, col)).map(|s| s.trim().to_string()) else {
            return Err(format!(
                "xlsx 헤더 {}열이 비어 있습니다 — 빈 헤더는 JSON으로 보내주세요",
                col + 1
            ));
        };
        if name.is_empty() {
            return Err(format!(
                "xlsx 헤더 {}열이 비어 있습니다 — 빈 헤더는 JSON으로 보내주세요",
                col + 1
            ));
        }
        headers.push(name);
    }

    let max_row = cells.keys().map(|(r, _)| *r).max().unwrap();
    let mut rows = Vec::new();
    for row in (min_row + 1)..=max_row {
        if let Some((_, col)) = cells.keys().find(|(r, c)| *r == row && *c > max_header) {
            return Err(format!(
                "xlsx {}행: 헤더 밖의 값이 있습니다 ({}열)",
                row,
                col + 1
            ));
        }
        let mut map = Map::new();
        let mut any = false;
        for (i, key) in headers.iter().enumerate() {
            let val = cells.get(&(row, i as u32)).cloned().unwrap_or_default();
            if !val.is_empty() {
                any = true;
            }
            map.insert(key.clone(), Value::String(val));
        }
        if any {
            rows.push(map);
        }
    }
    if rows.is_empty() {
        return Err("xlsx 데이터 행이 없습니다".into());
    }
    Ok(rows)
}

fn parse_cell_ref(r: &str) -> Option<(u32, u32)> {
    let r = r.trim_start_matches('$');
    let split = r.find(|c: char| c.is_ascii_digit())?;
    let (col_s, row_s) = r.split_at(split);
    let col_s = col_s.replace('$', "");
    let row: u32 = row_s.parse().ok()?;
    if row == 0 {
        return None;
    }
    Some((row, col_index(&col_s)?))
}

fn col_index(letters: &str) -> Option<u32> {
    if letters.is_empty() {
        return None;
    }
    let mut n: u32 = 0;
    for c in letters.chars() {
        let u = c.to_ascii_uppercase();
        if !u.is_ascii_uppercase() {
            return None;
        }
        n = n.checked_mul(26)?.checked_add(u as u32 - b'A' as u32 + 1)?;
    }
    n.checked_sub(1)
}

fn parse_merge(ref_: &str) -> Result<(u32, u32, u32, u32), String> {
    let (a, b) = ref_
        .split_once(':')
        .ok_or_else(|| format!("xlsx를 읽지 못했습니다: 병합 {ref_}"))?;
    let (r1, c1) =
        parse_cell_ref(a).ok_or_else(|| format!("xlsx를 읽지 못했습니다: 병합 {ref_}"))?;
    let (r2, c2) =
        parse_cell_ref(b).ok_or_else(|| format!("xlsx를 읽지 못했습니다: 병합 {ref_}"))?;
    Ok((r1, c1, r2, c2))
}
